kms/capi/capi.go

```diff
@@ -0,0 +1,646 @@
+	}
+
+	u, err := uri.ParseWithScheme(Scheme, req.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse URI: %w", err)
+	}
+
+	if k.containerName = u.Get(ContainerNameArg); k.containerName == "" {
+		// generate a uuid for the container name
+		k.containerName, err = randutil.UUIDv4()
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate uuid: %w", err)
+		}
+	}
+
+	alg, ok := signatureAlgorithmMapping[req.SignatureAlgorithm]
+	if !ok {
+		return nil, fmt.Errorf("unsupported algorithm %v", req.SignatureAlgorithm)
+	}
+
+	kh, err := nCryptCreatePersistedKey(k.providerHandle, k.containerName, alg, AT_KEYEXCHANGE, 0)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create persisted key")
+	}
+
+	defer nCryptFreeObject(kh)
+
+	if alg == "RSA" {
+		err = nCryptSetProperty(kh, NCRYPT_LENGTH_PROPERTY, uint32(req.Bits), 0)
+
+		if err != nil {
+			return nil, fmt.Errorf("unable to set key NCRYPT_LENGTH_PROPERTY: %w", err)
+		}
+	}
+
+	// users can store the key as a machine key by passing in storelocation = machine
+	// 'machine' is the only valid location, otherwise the key is stored as a 'user' key
+	storeLocation := u.Get(StoreLocationArg)
+
+	if storeLocation == "machine" {
+		err = nCryptSetProperty(kh, NCRYPT_KEY_TYPE_PROPERTY, NCRYPT_MACHINE_KEY_FLAG, 0)
+
+		if err != nil {
+			return nil, fmt.Errorf("unable to set key NCRYPT_KEY_TYPE_PROPERTY: %w", err)
+		}
+	} else if storeLocation != "" && storeLocation != "user" {
+		return nil, fmt.Errorf("invalid storeLocation %v", storeLocation)
+	}
+
+	// if supplied, set the smart card pin
+	scPIN := u.Pin()
+
+	//failover to pin set in kms instantiation
+	if scPIN == "" {
+		scPIN = k.pin
+	}
+
+	if scPIN != "" {
+		err = nCryptSetProperty(kh, NCRYPT_PIN_PROPERTY, scPIN, 0)
+
+		if err != nil {
+			return nil, fmt.Errorf("unable to set key NCRYPT_PIN_PROPERTY: %w", err)
+		}
+	}
+
+	err = nCryptFinalizeKey(kh, 0)
+	if err != nil {
+		return nil, fmt.Errorf("unable to finalize key: %w", err)
+	}
+
+	uc, err := nCryptGetPropertyStr(kh, NCRYPT_UNIQUE_NAME_PROPERTY)
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve NCRYPT_UNIQUE_NAME_PROPERTY: %w", err)
+	}
+
+	pub, err := getPublicKey(kh)
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve public key: %w", err)
+	}
+
+	return &apiv1.CreateKeyResponse{
+		Name:      uc,
+		PublicKey: pub,
+		CreateSignerRequest: apiv1.CreateSignerRequest{
+			SigningKey: uc,
+		},
+	}, nil
+}
+
+// GetPublicKey returns the public key from the file passed in the request name.
+func (k *CAPIKMS) GetPublicKey(req *apiv1.GetPublicKeyRequest) (crypto.PublicKey, error) {
+	var u *uri.URI
+
+	u, err := uri.ParseWithScheme(Scheme, req.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse URI: %w", err)
+	}
+
+	if k.containerName = u.Get(ContainerNameArg); k.containerName == "" {
+		return nil, fmt.Errorf("GetPublicKey %v not specified", ContainerNameArg)
+	}
```

Comment:
Do not overwrite containerName, see my previous comment. This should not be a field of CAPIKMS, it should always be in the URI.